File: doubly_linked_list.rb
require './node.rb'

class DoublyLinkedList
    def initalize()
        @head_node = nil
        @tail_node = nil
    end

    def add_to_head(new_value)
        new_head = Node.new(new_value)
        current_head = @head_node

        if !current_head.nil?
            current_head.prev_node = new_head
            new_head.next_node = current_head
        end

        @head_node = new_head

        if @tail_node.nil?
            @tail_node = new_head
        end
    end

    def add_to_tail(new_value)
        new_tail = Node.new(new_value)
        current_tail = @tail_node

        if !current_tail.nil?
            current_tail.next_node = new_tail
            new_tail.prev_node = current_tail
        end

        @tail_node = new_tail

        if @head_node.nil?
            @head_node = new_tail
        end
    end

    def remove_head()
        removed_head = @head_node

        if removed_head.nil?
            return nil
        end

        @head_node = removed_head.next_node

        if !@head_node.nil?
            @head_node.prev_node = nil
        end

        if removed_head == @tail_node
            self.remove_tail()
        end

        return removed_head.value 
    end

    def remove_tail()
        removed_tail = @tail_node

        if removed_tail.nil?
            return nil
        end

        @tail_node = removed_tail.prev_node

        if !@tail_node.nil?
            @tail_node.next_node = nil
        end

        if removed_tail == @head_node
            self.remove_head()
        end

        return removed_tail.value
    end

    def remove_by_value(value_to_remove)
        node_to_remove = nil
        current_node = @head_node

        while !current_node.nil?
            if current_node.value == value_to_remove
                node_to_remove = current_node
                break
            end
            current_node = current_node.next_node
        end

        if node_to_remove.nil?
            return nil
        end

        if node_to_remove == @head_node
            self.remove_head()
        elsif node_to_remove == @tail_node
            self.remove_tail()
        else
            next_node = node_to_remove.next_node
            prev_node = node_to_remove.prev_node
            next_node.prev_node = prev_node
            prev_node.next_node = next_node
        end

        return node_to_remove
    end

    def stringify_list()
        string = ""
        current_node = @head_node
        while current_node
            if not current_node.value.nil?
                string += current_node.value.to_s + "; "
            end
            current_node = current_node.next_node
        end
        return string
    end
end

File: DoublyLinkedList.cs
using System;
using System.Collections.Generic;

class Node<T>
{
    public T Value
    { get; private set; }
    public Node<T> NextNode
    { get; set; }
    public Node<T> PrevNode
    { get; set; }

    public Node(T value, Node<T> nextNode = null, Node<T> prevNode = null)
    {
        Value = value;
        NextNode = nextNode;
        PrevNode = prevNode;
    }
}


class DoublyLinkedList<T>
{
    public Node<T> HeadNode
    { get; private set; }
    public Node<T> TailNode
    { get; private set; }

    public DoublyLinkedList()
    {
        HeadNode = null;
        TailNode = null;
    }

    public void AddToHead(T newValue)
    {
        Node<T> newHead = new Node<T>(newValue);
        Node<T> currentHead = HeadNode;
        if (currentHead is not null)
        {
            currentHead.PrevNode = newHead;
            newHead.NextNode = currentHead;
        }

        HeadNode = newHead;

        if (TailNode is null)
        {
            TailNode = newHead;
        }
    }

    public void AddToTail(T newValue)
    {
        Node<T> newTail = new Node<T>(newValue);
        Node<T> currentTail = TailNode;
        if (currentTail is not null)
        {
            currentTail.NextNode = newTail;
            newTail.PrevNode = currentTail;
        }

        TailNode = newTail;

        if (HeadNode is null)
        {
            HeadNode = newTail;
        }
    }

    public T RemoveHead()
    {
        Node<T> removedHead = HeadNode;
        if (removedHead is null)
        {
            return default;
        }

        HeadNode = removedHead.NextNode;

        if (HeadNode is not null)
        {
            HeadNode.PrevNode = null;
        }

        if (removedHead == TailNode)
        {
            RemoveTail();
        }

        return removedHead.Value;
    }

    public T RemoveTail()
    {
        Node<T> removedTail = TailNode;
        if (removedTail is null)
        {
            return default;
        }

        TailNode = removedTail.PrevNode;

        if (TailNode is not null)
        {
            TailNode.NextNode = null;
        }

        if (removedTail == HeadNode)
        {
            RemoveHead();
        }

        return removedTail.Value;
    }

    public Node<T> RemoveByValue(T valueToRemove)
    {
        Node<T> nodeToRemove = null;
        Node<T> currentNode = HeadNode;
        while (currentNode is not null)
        {
            if (EqualityComparer<T>.Default.Equals(currentNode.Value, valueToRemove))
            {
                nodeToRemove = currentNode;
                break;
            }
            currentNode = currentNode.NextNode;
        }

        if (nodeToRemove == null)
        {
            return default;
        }

        if (EqualityComparer<Node<T>>.Default.Equals(nodeToRemove, HeadNode))
        {
            RemoveHead();
        }
        else if (nodeToRemove == TailNode)
        {
            RemoveTail();
        }
        else
        {
            Node<T> next = nodeToRemove.NextNode;
            Node<T> prev = nodeToRemove.PrevNode;
            next.PrevNode = prev;
            prev.NextNode = next;
        }

        return nodeToRemove;
    }

        public string Stringify()
        {
            string s = "";
            Node<T> current = HeadNode;
            while (current != null)
            {
                if (current.Value != null)
                {
                    s += Convert.ToString(current.Value ) + "; ";
                }
                current = current.NextNode;
            }
            return s;
        }
}